File: Day 3: Arrays Part-III/Search in 2D matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m=matrix.size(),n=matrix[0].size();
        int low = 0, high=(m*n)-1;
        while(low<=high){
            int mid = low+((high-low)>>1);
            int x = mid/n, y=mid%n;
            if(matrix[x][y]==target) return true;
            else if (matrix[x][y]>target) high=mid-1;
            else low=mid+1;
        }
        return false;
    }
};

//  O(log(m*n))-time
//  O(1)-space

File: Day 3: Arrays Part-III/Java/Search-In-2D-Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length, m = matrix[0].length;
        int l = 0, h = (m*n)-1;
        int t = Math.max(m,n);
        while(l<=h){
            int mid = l+((h-l)>>1);
            int x = mid/m, y = mid%m;
            if(matrix[x][y]==target) return true;
            else if(matrix[x][y]>target) h = mid-1;
            else l = mid+1;
        }
        return false;
    }
}

//O(NlogN) time
